// @1inch/fusion-sdk/dist/esm/sdk/sdk.js
import { Address, MakerTraits } from '@1inch/limit-order-sdk';
import { encodeCancelOrder } from './encoders';
import { FusionApi, QuoterRequest, RelayerRequest, QuoterCustomPresetRequest } from '../api';
import { ActiveOrdersRequest, OrdersByMakerRequest, OrderStatusRequest } from '../api/orders';
export class FusionSDK {
    constructor(config) {
        this.config = config;
        this.api = FusionApi.new({
            url: config.url,
            network: config.network,
            httpProvider: config.httpProvider,
            authKey: config.authKey
        });
    }
    async getActiveOrders({ page, limit } = {}) {
        const request = ActiveOrdersRequest.new({ page, limit });
        return this.api.getActiveOrders(request);
    }
    async getOrderStatus(orderHash) {
        const request = OrderStatusRequest.new({ orderHash });
        return this.api.getOrderStatus(request);
    }
    async getOrdersByMaker({ limit, page, address }) {
        const request = OrdersByMakerRequest.new({ limit, page, address });
        return this.api.getOrdersByMaker(request);
    }
    async getQuote(params) {
        const request = QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: params.walletAddress || Address.ZERO_ADDRESS.toString(),
            permit: params.permit,
            enableEstimate: !!params.enableEstimate,
            fee: params?.takingFeeBps,
            source: params.source,
            isPermit2: params.isPermit2
        });
        return this.api.getQuote(request);
    }
    async getQuoteWithCustomPreset(params, body) {
        const paramsRequest = QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: params.walletAddress || Address.ZERO_ADDRESS.toString(),
            permit: params.permit,
            enableEstimate: !!params.enableEstimate,
            fee: params?.takingFeeBps,
            source: params.source,
            isPermit2: params.isPermit2
        });
        const bodyRequest = QuoterCustomPresetRequest.new({
            customPreset: body.customPreset
        });
        return this.api.getQuoteWithCustomPreset(paramsRequest, bodyRequest);
    }
    async createOrder(params) {
        const quote = await this.getQuoteResult(params);
        if (!quote.quoteId) {
            throw new Error('quoter has not returned quoteId');
        }
        const order = quote.createFusionOrder({
            receiver: params.receiver
                ? new Address(params.receiver)
                : undefined,
            preset: params.preset,
            nonce: params.nonce,
            takingFeeReceiver: params.fee?.takingFeeReceiver,
            allowPartialFills: params.allowPartialFills,
            allowMultipleFills: params.allowMultipleFills,
            orderExpirationDelay: params.orderExpirationDelay,
            network: this.config.network
        });
        const hash = order.getOrderHash(this.config.network);
        return { order, hash, quoteId: quote.quoteId };
    }
    async submitOrder(order, quoteId) {
        if (!this.config.blockchainProvider) {
            throw new Error('blockchainProvider has not set to config');
        }
        const orderStruct = order.build();
        const signature = await this.config.blockchainProvider.signTypedData(orderStruct.maker, order.getTypedData(this.config.network));
        const relayerRequest = RelayerRequest.new({
            order: orderStruct,
            signature,
            quoteId,
            extension: order.extension.encode()
        });
        await this.api.submitOrder(relayerRequest);
        return {
            order: orderStruct,
            signature,
            quoteId,
            orderHash: order.getOrderHash(this.config.network),
            extension: relayerRequest.extension
        };
    }
    async placeOrder(params) {
        const { order, quoteId } = await this.createOrder(params);
        return this.submitOrder(order, quoteId);
    }
    async buildCancelOrderCallData(orderHash) {
        const getOrderRequest = OrderStatusRequest.new({ orderHash });
        const orderData = await this.api.getOrderStatus(getOrderRequest);
        if (!orderData) {
            throw new Error(`Can not get order with the specified orderHash ${orderHash}`);
        }
        const { order } = orderData;
        return encodeCancelOrder(orderHash, new MakerTraits(BigInt(order.makerTraits)));
    }
    async signOrder(order) {
        if (!this.config.blockchainProvider) {
            throw new Error('blockchainProvider has not set to config');
        }
        const orderStruct = order.build();
        const data = order.getTypedData(this.config.network);
        return this.config.blockchainProvider.signTypedData(orderStruct.maker, data);
    }
    async getQuoteResult(params) {
        const quoterRequest = QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: params.walletAddress,
            permit: params.permit,
            enableEstimate: true,
            fee: params.fee?.takingFeeBps,
            source: params.source,
            isPermit2: params.isPermit2
        });
        if (!params.customPreset) {
            return this.api.getQuote(quoterRequest);
        }
        const quoterWithCustomPresetBodyRequest = QuoterCustomPresetRequest.new({
            customPreset: params.customPreset
        });
        return this.api.getQuoteWithCustomPreset(quoterRequest, quoterWithCustomPresetBodyRequest);
    }
}
