// @1inch/cross-chain-sdk/dist/cjs/ws-api/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketEvent = exports.RpcMethod = exports.EventType = void 0;
var EventType;
(function (EventType) {
    EventType["OrderCreated"] = "order_created";
    EventType["OrderInvalid"] = "order_invalid";
    EventType["OrderBalanceChange"] = "order_balance_change";
    EventType["OrderAllowanceChange"] = "order_allowance_change";
    EventType["OrderFilled"] = "order_filled";
    EventType["OrderFilledPartially"] = "order_filled_partially";
    EventType["OrderCancelled"] = "order_cancelled";
    EventType["OrderSecretShared"] = "secret_shared";
})(EventType || (exports.EventType = EventType = {}));
var RpcMethod;
(function (RpcMethod) {
    RpcMethod["GetAllowedMethods"] = "getAllowedMethods";
    RpcMethod["Ping"] = "ping";
    RpcMethod["GetActiveOrders"] = "getActiveOrders";
    RpcMethod["GetSecrets"] = "getSecrets";
})(RpcMethod || (exports.RpcMethod = RpcMethod = {}));
var WebSocketEvent;
(function (WebSocketEvent) {
    WebSocketEvent["Close"] = "close";
    WebSocketEvent["Error"] = "error";
    WebSocketEvent["Message"] = "message";
    WebSocketEvent["Open"] = "open";
})(WebSocketEvent || (exports.WebSocketEvent = WebSocketEvent = {}));
